### DISCUSSION TITLE
How to make allreduce and all2all run in parallel?

### DISCUSSION THREAD

**Discussion by zhuyijie** [horovod/horovod#2784] [Q&A]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 2.4.0
3. Horovod version: 0.21.3

**Your question:**
Hi, I have an wide&deep model which use all2all to handle sparse vars and allreduce for dense vars. I've observed that the all2all and allreduce are mutually exclusive, please see the profile below. I'm wondering if I can make all2all and allreduce to run in parallel? or can I make multiple all2all calls to run in parallel?

<img width="1436" alt="Screen Shot 2021-02-18 at 11 28 40 AM" src="https://user-images.githubusercontent.com/2706249/108410612-7e430f80-71dc-11eb-8680-59f3ec1b4537.png">

**Comment by tgaddair** [comment]
Hey @zhuyijie, this is an interesting observation. @romerojosh would something like this be possible using multiple NCCL streams?

**Comment by zhuyijie** [comment]
> Hey @zhuyijie, this is an interesting observation. @romerojosh would something like this be possible using multiple NCCL streams?

Hi, @tgaddair, Thank you replying! Just want to clarify, we are not using nccl. We are using mpi to run sync training on CPUs

**Comment by maxhgerlach** [comment]
Would the enhancement discussed in #2139 cover this use case? It looks as if only a subset of the ranks are involved in the allreduce, but block the other ranks waiting for the alltoall.
